### ISSUE TITLE
Fritz!Box 6591 Cable UPnPError 401 invalid action

### ISSUE THREAD

**Issue by tzockt** [pdreker/fritzbox_exporter#25]
Hi,

Currently, having an issue with the script, and the error is not really helping:
I have a FritzBox 6591 Cable with FRITZ!OS 07.13

First a quick description of the libs:
```
root@bababooey:/opt/fritzbox_exporter# pip3 install prometheus-client
Requirement already satisfied: prometheus-client in /usr/local/lib/python3.8/dist-packages (0.9.0)
root@bababooey:/opt/fritzbox_exporter# pip3 install fritzconnection
Requirement already satisfied: fritzconnection in /usr/local/lib/python3.8/dist-packages (1.4.1)
Requirement already satisfied: requests>=2.22.0 in /usr/lib/python3/dist-packages (from fritzconnection) (2.22.0)
```
When I execute the Script:
````
root@bababooey:/opt/fritzbox_exporter# python3 fritz_exporter.py
Traceback (most recent call last):
  File "fritz_exporter.py", line 59, in <module>
    main()
  File "fritz_exporter.py", line 46, in main
    REGISTRY.register(fritzcollector)
  File "/usr/local/lib/python3.8/dist-packages/prometheus_client/registry.py", line 26, in register
    names = self._get_names(collector)
  File "/usr/local/lib/python3.8/dist-packages/prometheus_client/registry.py", line 66, in _get_names
    for metric in desc_func():
  File "/opt/fritzbox_exporter/fritzexporter/fritzdevice.py", line 74, in collect
    yield from capa.getMetrics(self.devices, name)
  File "/opt/fritzbox_exporter/fritzexporter/fritzcapabilities.py", line 28, in getMetrics
    yield from self._getMetricValues(device)
  File "/opt/fritzbox_exporter/fritzexporter/fritzcapabilities.py", line 202, in _getMetricValues
    fritz_dslinfo_result = device.fc.call_action('WANDSLInterfaceConfig:1', 'GetInfo')
  File "/usr/local/lib/python3.8/dist-packages/fritzconnection/core/fritzconnection.py", line 227, in call_action
    return self.soaper.execute(service, action_name, arguments)
  File "/usr/local/lib/python3.8/dist-packages/fritzconnection/core/soaper.py", line 238, in execute
    return handle_response(response)
  File "/usr/local/lib/python3.8/dist-packages/fritzconnection/core/soaper.py", line 222, in handle_response
    raise_fritzconnection_error(response)
  File "/usr/local/lib/python3.8/dist-packages/fritzconnection/core/soaper.py", line 147, in raise_fritzconnection_error
    raise exception(message)
fritzconnection.core.exceptions.FritzActionError: UPnPError:
errorCode: 401
errorDescription: Invalid Action
````

**Comment by pdreker** [comment]
At first glance it looks like the capability detection went wrong, and your box seems to report DSL stats as available but then responds with "Invalid Action" when actually trying to query the stats.

As you're the "first cable customer" I've had I'd like to see some output from fritzconnection (Command line utility delivered with the fritzconnection python module).

I'll be back at my computer on Sunday, so I can provide more detailed instructions.

**Comment by pdreker** [comment]
Ok, I managed to find the commands using my cellphone...

Could you run the following commands and provide me with the output (preferable using https://gist.github.com/)?

```
fritzconnection -i FRITZBOXIP -s 

fritzconnection -i FRITZBOXIP -S WANDSLInterfaceConfig1
```

**Comment by tzockt** [comment]
Here is the Output:
https://gist.github.com/tzockt/bdaa35286c88fe4ae2915b03d5c70a98

**Comment by pdreker** [comment]
Indeed, somewhat like I expected. The box reports WanDslInterfaceConfig:GetInfo as available, but responds with "Invalid Action" when actually calling it.

I will have to come up with some code for you to test, to figure out what to do here.

I'll get back to you on Sunday.

**Comment by pdreker** [comment]
I have just pushed a new branch to the repository, which has some code to test. There a actually 2 things in there:

1. the exporter it self should be more conservative as to which actions are actually workable on a box. It will still not report the metrics, but it should actually just log a warning, that an action was disabled and continue running, instead of throwing the exception and exiting.
2. I have made a small helper script to test out actions on the API. The script can be run with `python3 -m fritz_export_helper FRITZ_IP USERNAME PASSWORD SERVICE ACTION`.

To get to the new code just clone the git repo and run `git checkout fix-cable-boxes`. The code should then be visible in the source folder.

I would need two things: First, check if the exporter now keeps running, instead of erroring out. WAN Statistics will be missing, but it should run (Fingers crossed...). Second: to actually report useful cable stats I will need to see, what is actually reported back from a cable box. That's what the helper script is for.

Could you please check some services with `fritzconnection -i FRITZBOXIP -S SERVICENAME`. This will report back the actions supported under that service. Normally there is a "GetInfo" or "GetStats" action in there somewhere, but it may actually be called differently. Find those "interesting" service/action combos and run the helper script against then and let me see the output (plus of course which service/action was called). Check that the output does not contain passwords or other secret data!

From your first reply, I cannot see, if there is a specific category for the cable uplink (for DSL Boxes it is WANDSLInterfaceConfig, which errors on your box). So you could maybe check that explicit service again with the helper script, if it still delivers an error. Otherwise just have a look around the Service/Action pairs reported by the box. If you find something which is missing from the Prometheus output, let me know.

Also: the 401 error from the box *can* mean "Unauthorized" - is your password correct? I had another issue in the past, where the password used was too long. Even though docs state the password can be 32 characters long I have found the only 30 or 31 chars actually work. Is the box managed by your provider or do you have full access? If you do not have full access, this could also limit the API but that's pure speculation from my side...

Thanks!

**Comment by tzockt** [comment]
Hi,

Do you have Discord if so add me TZockt#4655.
Then we can solve this directly because I do not understand what exactly you need. Likewise, we can then gladly write in German :)

**Comment by pdreker** [comment]
I just sent you a discord request. Your first reply came so quick, that I just assumed, you'd know, what I wanted ;-)

I'll be "on and off" during the evening, so just watch out for me (goofy42#....)

**Comment by pdreker** [comment]
I have released 1.0.1 with a fix, which should enable cable boxes to at least run without an error. There will be warnings logged, which will also show the error, but the exporter will not crash due to the unhandled exception.
